src/byte_array.rs

```diff
@@ -0,0 +1,264 @@
+// Copyright 2019, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+//! # Examples
+//!
+//! ```
+//! use glib::prelude::*; // or `use gtk::prelude::*;`
+//! use glib::ByteArray;
+//!
+//! let ba = ByteArray::from(b"def");
+//! ba.append(b"ghi").prepend(b"abc");
+//! ba.remove_range(3, 3);
+//! assert_eq!(ba, "abcghi".as_bytes());
+//! ```
+
+use glib_sys;
+use std::borrow::Borrow;
+use std::cmp::Ordering;
+use std::fmt;
+use std::hash::{Hash, Hasher};
+use std::mem;
+use std::ops::Deref;
+use std::slice;
+use translate::*;
+
+use Bytes;
+
+glib_wrapper! {
+    pub struct ByteArray(Shared<glib_sys::GByteArray>);
+
+    match fn {
+        ref => |ptr| glib_sys::g_byte_array_ref(ptr),
+        unref => |ptr| glib_sys::g_byte_array_unref(ptr),
+        get_type => || glib_sys::g_byte_array_get_type(),
+    }
+}
+
+impl ByteArray {
+    pub fn new() -> ByteArray {
+        unsafe { from_glib_full(glib_sys::g_byte_array_new()) }
+    }
+
+    pub fn with_capacity(size: usize) -> ByteArray {
+        unsafe { from_glib_full(glib_sys::g_byte_array_sized_new(size as u32)) }
+    }
+
+    pub fn into_gbytes(self) -> Bytes {
+        unsafe {
+            let ret = from_glib_full(glib_sys::g_byte_array_free_to_bytes(mut_override(
+                self.to_glib_none().0,
+            )));
+            mem::forget(self);
+            ret
+        }
+    }
+
+    pub fn append<T: AsRef<[u8]>>(&self, data: T) -> &Self {
+        let bytes = data.as_ref();
+        unsafe {
+            glib_sys::g_byte_array_append(
+                self.to_glib_none().0,
+                bytes.as_ptr() as *const _,
+                bytes.len() as u32,
+            );
+        }
+        self
+    }
+
+    pub fn prepend<T: AsRef<[u8]>>(&self, data: T) -> &Self {
```

Comment:
You could here (and above) take the data by reference instead. There's no need to take ownership of it here as the byte array will copy the data